### ISSUE TITLE
incorrect special_tokens_mask

### ISSUE THREAD

**Issue by MostHumble** [huggingface/transformers#35897]
### System Info

- `transformers` version: 4.47.1
- Platform: Linux-6.1.85+-x86_64-with-glibc2.35
- Python version: 3.11.11
- Huggingface_hub version: 0.27.1
- Safetensors version: 0.5.2
- Accelerate version: 1.2.1
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.5.1+cu121 (True)
- Tensorflow version (GPU?): 2.17.1 (True)
- Flax version (CPU?/GPU?/TPU?): 0.10.2 (gpu)
- Jax version: 0.4.33
- JaxLib version: 0.4.33
- Using distributed or parallel set-up in script?: False
- Using GPU in script?: False
- GPU type: Tesla T4

### Who can help?

@ArthurZucker @itazap 

### Information

- [x] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
checkpoint = "HuggingFaceTB/SmolLM2-135M" # same behavior with gpt2
device = "cuda" # for GPU usage or "cpu" for CPU usage
tokenizer = AutoTokenizer.from_pretrained(checkpoint)
model = AutoModelForCausalLM.from_pretrained(checkpoint).to(device)
outputs = tokenizer("<|endoftext|><|im_start|>y", return_tensors="pt", return_special_tokens_mask=True)
```
outputs:
```cmd
{'input_ids': tensor([[  0,   1, 105]]), ..., 'special_tokens_mask': tensor([[0, 0, 0]])}
```

### Expected behavior

outputs:
```cmd
{'input_ids': tensor([[  0,   1, 105]]), ..., 'special_tokens_mask': tensor([[1, 1, 0]])}
```

given that:

```python
tokenizer.special_tokens_map
```

```cmd
{'bos_token': '<|endoftext|>',
 'eos_token': '<|endoftext|>',
 'unk_token': '<|endoftext|>',
 'additional_special_tokens': ['<|endoftext|>',
  '<|im_start|>',
  '<|im_end|>',
...
```

**Comment by Rocketknight1** [comment]
The tokenizer code is written to only return that mask correctly for tokens that it added itself (i.e. when `add_special_tokens=True`). I'm not sure it's designed to handle cases when the user has encoded special tokens in the prompt itself, see e.g. [here](https://github.com/huggingface/transformers/blob/main/src/transformers/tokenization_utils_base.py#L3544)

cc @ArthurZucker for tokenizers - this is intended behaviour and not a bug, right?

**Comment by MostHumble** [comment]
AFAIK they choose by design to have ```add_special_tokens=True``` not do anything for this tokenizer ([#35862](https://github.com/huggingface/transformers/issues/35862)) . that's actually why I ended asking this.
